metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java

```diff
@@ -319,12 +327,51 @@ private boolean isGroupOwner(Set<Urn> groups, Set<String> owners) {
     return groups.stream().anyMatch(group -> owners.contains(group.toString()));
   }
 
+  private boolean isRoleMatch(final Urn actor, final DataHubActorFilter actorFilter,
+      final PolicyEvaluationContext context) {
+    // Can immediately return false if the actor filter does not have any roles
+    if (!actorFilter.hasRoles()) {
+      return false;
+    }
+    // If the actor has a matching "Role" in the actor filter, return true immediately.
+    Set<Urn> actorRoles = resolveRoles(actor, context);
+    return Objects.requireNonNull(actorFilter.getRoles())
+        .stream()
+        .anyMatch(actorRoles::contains);
+  }
+
+  private Set<Urn> resolveRoles(Urn actor, PolicyEvaluationContext context) {
+    if (context.roles != null) {
+      return context.roles;
+    }
+
+    try {
+      final EntityResponse corpUser =
+          _entityClient.batchGetV2(CORP_USER_ENTITY_NAME, Collections.singleton(actor), null, _systemAuthentication)
```

Comment:
Whoa whoa whoa --- We cannot use "null" as the third argument here. That will fetch _ALL_ aspects for the user. This can become expensive as we add more aspects.

Please use this instead:

ImmutableSet.of(CORP_USER_ROLES_ASPECT)

Reply:
Refactored!